ta/pkcs11/src/processing.c

```diff
@@ -631,3 +631,35 @@ enum pkcs11_rc entry_derive_key(struct pkcs11_client *client,
 
 	return rc;
 }
+
+enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+					       uint32_t ptypes,
+					       TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	release_active_processing(session);
```

Comment:
Should we get the target processing (Digest/encrypt/Wrap/...) from an argument?

If during an on-going Digest one calls C_Encrypt() with an invalid NULL/NULL pbuf/psize reference, should the client library request to terminate the Digest? Only NULL/NULL for C_Digest*() should terminate the active processing. Other C_xxx should get OPERATION_NOT_INITIALIZED.

I don't think there are security matters here but I would prefer we don't introduce unexpected behavior from the client lib.

If the client lib passes the CMD_xxx value for the C_xxx() function called, next to the session handle, we we could get the related PKCS11_FUNCTION_xxx and call `get_active_session(session, function);` before `release_active_processing(session);`.

What do you think. It is worth it?
In this P-R or and a later one?

Reply:
Sounds reasonable.

Reply:
So I suppose we would need helper to determine function based on `enum pkcs11_ta_cmd` which would then return `enum processing_func`. Let's see what kind of code that would be.

Reply:
Added support in the PR.